allocation: decline allocation when node is hotspotting

schase-es opened the pull request:
After a node that is write-load hotspotting moves shards away to another node, it has a period of 30 seconds (until the next ClusterInfo) where the queue latency is simulated as 0. This is to prevent more shards from being relocated away, until a new ClusterInfo shows whether the latency has been resolved. This change adds a flag set whenever the node is hotspotting and its queue latency is being simulated as 0. The write load decider checks this flag, and returns not preferred to prevent shards from being moved onto the node during this interim.

schase-es committed:
allocation: decline allocation when node is hotspotting

After a node that is write-load hotspotting moves shards away to another node,
it has a period of 30 seconds (until the next ClusterInfo) where the queue
latency is simulated as 0. This is to prevent more shards from being relocated
away, until a new ClusterInfo shows whether the latency has been resolved. This
change adds a flag set whenever the node is hotspotting and its queue latency is
being simulated as 0. The write load decider checks this flag, and returns not
preferred to prevent shards from being moved onto the node during this interim.

Commit:
[CI] Auto commit changes from spotless

schase-es committed:
Fixing up confusion with canAllocate and canRemain

DiannaHohensee commented on server/src/main/java/org/elasticsearch/cluster/routing/ShardMovementWriteLoadSimulator.java:
```diff
@@ -107,7 +107,8 @@ private static NodeUsageStatsForThreadPools.ThreadPoolUsageStats replaceWritePoo
             ),
-            adjustThreadPoolQueueLatencyWithShardMovements(writeThreadPoolStats.maxThreadPoolQueueLatencyMillis(), hasSeenMovedAwayShard)
+            adjustThreadPoolQueueLatencyWithShardMovements(writeThreadPoolStats.maxThreadPoolQueueLatencyMillis(), hasSeenMovedAwayShard),
+            hasSeenMovedAwayShard
```
`hasSeenMovedAwayShard` isn't set based on there being a hot-spot or not, IIUC. So this this would set isHotspotting [whenever a shard moves away from a node](https://github.com/elastic/elasticsearch/blob/1b26cc5a1b33c876a2567644d09c67e528c2082d/server/src/main/java/org/elasticsearch/cluster/routing/ShardMovementWriteLoadSimulator.java#L61-L64)?

schase-es replied:
That is what I thought -- but this also simulates the queue latency as 0. Does having the latency as 0 have an impact? I suppose it prevents it from being considered hotspotting for the next stretch.

The right thing to do is set this flag whenever the write load decider flags something as a hotspot.

schase-es replied:
I implemented one of Dianna's suggestions, which is to do a threshold calculation during the transition from the cluster info simulator to cluster info.

Another approach is to do the threshold calculation during the node action.

schase-es committed:
Set isHotspotting flag whenever original data is above threshold

schase-es committed:
Adding tests

DiannaHohensee commented on server/src/main/java/org/elasticsearch/cluster/routing/ShardMovementWriteLoadSimulator.java:
ClusterInfoSimulator#getClusterInfo is the only caller of this method and is called often -- once per shard move in a computation round. Whether a node has a hot-spot, however, never changes for the lifetime of a simulator (a single computation round). It would be better to compute what nodes are hot-spotting up front, and avoid the cost. The constructor would work for that?

DiannaHohensee reviewed (commented):
Left some high level notes.

schase-es commented on server/src/main/java/org/elasticsearch/cluster/routing/ShardMovementWriteLoadSimulator.java:
```diff
@@ -81,11 +91,22 @@ public Map<String, NodeUsageStatsForThreadPools> simulatedNodeUsageStatsForThrea
+                    Maps.copyMapWithAddedOrReplacedEntry(
+                        entry.getValue().threadPoolUsageStatsMap(),
+                        ThreadPool.Names.WRITE,
+                        setIsHotspottingFlag(entry.getValue())
```
The hotspotting flag actually isn't included in the serialization. It's implemented as an empty default, and set in the simulator for use in the allocation code. I was trying to implement this efficiently, but reusing the same structure for data node and aggregate use is confusing. This breaks a common pattern, and is a good reason to move it somewhere else. I'll create a separate table.

I'm mentioning the immutability of `NodeUsageStatsForThreadPools` because of the intensiveness of how the simulator recomputes it. In `ShardMovementWriteLoadSimulator`, the `simulatedNodeUsageStatsForThreadPools` method is already inefficient in how it creates the map: it rebuilds the map with each call, and recreates the simulated entries for relevant `NodeUsageStatsForThreadPools`.

I specifically wasn't concerned about the performance of the code I added, because it was already inefficient. If we do turn out to care about this, we should move towards a mutable aggregate structure.

DiannaHohensee replied:
Thanks for the map 👍 

> The hotspotting flag actually isn't included in the serialization. It's implemented as an empty default, and set in the simulator for use in the allocation code.

Ah. I hadn't realized the flag wasn't added to the serialization. Fair point. But yeah, the purpose of `NodeUsageStatsForThreadPools` is to hold what we send over the wire.

> I specifically wasn't concerned about the performance of the code I added, because it was already inefficient.

Yep, agreed. It's a reason to avoid adding to it, in my mind, since fixing / investigating it is out of scope right now. I know Nick pointed out on slack that the logic is called relatively little on average, but it's good practice to avoid potential problems by default (who know what happens to the code in future), and there are always going to be the non-average use case outliers.

schase-es committed:
Reverting NodeUsageStatsForThreadPools implementation

schase-es committed:
Reverting simulator test

schase-es committed:
Implemented hotspotting flag as cluster info table

Commit:
[CI] Auto commit changes from spotless

DiannaHohensee commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderUnitTests.java:
optional: there's technically a ClusterInfo#Builder that could replace this usage, so we won't have to update it in future. E.g.

```
        final ClusterInfo clusterInfo = ClusterInfo.builder()
            .leastAvailableSpaceUsage(leastAvailableUsages)
            .mostAvailableSpaceUsage(mostAvailableUsage)
            .shardSizes(Map.of("[test][0][p]", 10L))
            .build();
```

DiannaHohensee commented on server/src/main/java/org/elasticsearch/cluster/ClusterInfo.java:
```diff
@@ -639,5 +681,14 @@ public Builder maxHeapSizePerNode(Map<String, ByteSizeValue> maxHeapSizePerNode)
             return this;
         }
+
+        public Builder nodesWriteLoadHotspotting(Map<String, Boolean> nodesWriteLoadHotspotting) {
```
We typically tag test only methods with a comment, e.g.
```suggestion
        // Visible for testing
        public Builder nodesWriteLoadHotspotting(Map<String, Boolean> nodesWriteLoadHotspotting) {
```

DiannaHohensee commented on server/src/main/java/org/elasticsearch/cluster/ClusterInfo.java:
```diff
@@ -72,10 +76,11 @@ public class ClusterInfo implements ChunkedToXContent, Writeable, ExpectedShardS
     final Map<ShardId, Double> shardWriteLoads;
     // max heap size per node ID
     final Map<String, ByteSizeValue> maxHeapSizePerNode;
+    final Map<String, Boolean> nodesWriteLoadHotspotting;
```
Could you add either ID in the variable name or a comment mentioning it? Easier than looking around to figure out whether we've got a map by node name or node id.

DiannaHohensee commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/WriteLoadConstraintDeciderTests.java:
```diff
@@ -369,6 +371,77 @@ public void testWriteLoadDeciderShouldPreventBalancerMovingShardsBack() {
+            )
+            .build();
+
+        final var state = ClusterStateCreationUtils.state(2, new String[] { indexName }, 4);
```
Looks like you only need 1 node here, not two? Should be in sync with the ClusterInfo.

DiannaHohensee reviewed (commented):
Approach looks good, leaving some tidying up comments.

Could you move this out of draft mode? I don't think I'll be able to approve it properly otherwise.

nicktindall reviewed (approved):
This LGTM pending Dianna's approval.

schase-es committed:
Renaming field

schase-es committed:
Move of hotspot flag table out of cluster info

schase-es committed:
random generation test

schase-es committed:
Compilation fixes

schase-es committed:
Add all nodes to node thread pool map

schase-es committed:
Use utility constructor for write load decider

schase-es committed:
Logging fix for hotspot status

schase-es committed:
Use writeBoolean in map serialization

Commit:
[CI] Auto commit changes from spotless

schase-es committed:
Make node ids a set of those hotspotting

schase-es committed:
Fixup log message

nicktindall reviewed (approved):
This LGTM

DiannaHohensee reviewed (approved):
LGTM 👍 I left a few small requests.

In future, please move out of DRAFT mode if you're seeking a lgtm, and please respond to the reviewer comments -- however you'd like to respond, can be a conversation, can be a counter-argument, or as simple as "Done" or a thumbs-up emoticon -- so the reviewer is aware the comments have all been addressed.

schase-es committed:
Time value pass-through and test fixup

schase-es commented on server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java:
```diff
@@ -479,11 +487,27 @@ private ClusterInfo updateAndGetCurrentClusterInfo() {
+
+        public static Set<String> buildNodeIdsWriteLoadHotspottingSet(
+            Map<String, NodeUsageStatsForThreadPools> nodeThreadPoolUsageStatsPerNode,
+            long queueLatencyThreshold
```
Done

schase-es committed:
Merge branch 'main' into ES-13549-write-load-hotspot-flag-prevents-allocation

Commit:
[CI] Update transport version definitions

schase-es committed:
Refactor to share a single WriteLoadConstraintSettings object

InternalClusterInfoService and WriteLoadConstraintMonitor made up their own
WriteLoadConstraintSettings object from ClusterSettings. This refactors the node
construction and node services and their tests to use one, passed in from node
construction.

schase-es committed:
Merge branch 'main' into ES-13549-write-load-hotspot-flag-prevents-allocation

Commit:
[CI] Update transport version definitions

nicktindall reviewed (approved):
LGTM

schase-es committed:
Merge branch 'main' into ES-13549-write-load-hotspot-flag-prevents-allocation

Commit:
[CI] Update transport version definitions

schase-es committed:
Merge branch 'main' into ES-13549-write-load-hotspot-flag-prevents-allocation

schase-es committed:
Test creation fixup

schase-es committed:
Merge branch 'main' into ES-13549-write-load-hotspot-flag-prevents-allocation

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

Commit:
[CI] Update transport version definitions

schase-es committed:
Fixup upper bounds

Commit:
[CI] Update transport version definitions
